[GDScript] Add refactor rename symbol functionality

adamscott opened the pull request:
> [!WARNING]
> ~~For this moment, this PR is public for testing purposes. Work is still done to it to iron out issues.~~
> ~~Use at your own risk AND always backup before trying out new features.~~

This PR adds the refactor rename symbol functionality to the GDScript code editor inside the Godot editor.

To use this feature, just click on the desired symbol to rename and press F2 on your keyboard. You can also use the contextual menu (right click) or use the entry in the Edit menu.

![Capture d’écran du 2025-02-03 17-15-43](https://github.com/user-attachments/assets/360be407-bf3b-41b4-835c-1347b1b7d6bf)

Fixes https://github.com/godotengine/godot-proposals/issues/899

adamscott committed:
Add refactor rename symbol functionality

ZanyaNova reviewed (commented):
Really hoping this becomes a thing! I'm terrible with naming things correctly on my first pass.

HolonProduction commented:
- When using this in an unsaved file the changes get mixed up with the code from disk.
- Doesn't seem to work for constants and signals at all at the moment
- Doesn't pick up usages as type hint:
```gdscript
class Test:
	pass

var with_type: Test
```
- The following scenario with two scripts:
```gdscript
# a.gd
extends Node
class_name Test1

const SIZE_EXPAND = 1

static var USUAL = 2
```
```gdscript
# b.gd
extends Node

func ready(delta: float) -> void:
	Test1.SIZE_EXPAND # Renaming SIZE_EXPAND will modify b.gd but not a.gd
	Test1.USUAL # Renaming USUAL will modify a.gd but not b.gd
	pass
```
Without having looked at the implementation: Collecting all usages and replacing them needs to happen in two passes, or the change might destroy references that can not be found anymore.
- Shadowing:
```
extends Node

var delta: float # Renaming delta will rename the parameter name as well but not the usages
func test(delta: float): # Renaming delta will change the occurence in print but not the standalone occurence
	print(delta) # Renaming delta will change the parameter name but not the standalone occurence
	delta # Renaming delta will change the parameter name and the occurence in print but not the standalone occurence

```

adamscott commented:
> Could a 'Match Case' functionality be considered for this? Something like renaming `test` to `test_renamed` also changes `TEST` to `TEST_RENAMED`, `Test` to `TestRenamed`, etc.?

I don't understand.

BenjaTK commented:
> > Could a 'Match Case' functionality be considered for this? Something like renaming `test` to `test_renamed` also changes `TEST` to `TEST_RENAMED`, `Test` to `TestRenamed`, etc.?
> 
> I don't understand.

See IntelliJ IDEA for example.

If you rename the class `MyClass` to `YourClass`:
#### Before
```gdscript
class_name MyClass

[...]

my_class = MyClass.new()
```

#### After
```gdscript
class_name YourClass

[...]

your_class = YourClass.new()
```

adamscott commented:
> > > Could a 'Match Case' functionality be considered for this? Something like renaming `test` to `test_renamed` also changes `TEST` to `TEST_RENAMED`, `Test` to `TestRenamed`, etc.?
> > 
> > 
> > I don't understand.
> 
> See IntelliJ IDEA for example.
> 
> If you rename the class `MyClass` to `YourClass`:
> #### Before
> 
> ```gdscript
> class_name MyClass
> 
> [...]
> 
> var my_class = MyClass.new()
> ```
> 
> #### After
> 
> ```gdscript
> class_name YourClass
> 
> [...]
> 
> var your_class = YourClass.new()
> ```

It's not in the current scope of the PR.

Calinou reviewed (commented):
Tested locally on the [3D platformer demo](https://github.com/godotengine/godot-demo-projects/tree/master/3d/platformer), it mostly works as expected.

Some feedback:

- If you select a word by double-clicking or <kbd>Ctrl + D</kbd> then try to use <kbd>F2</kbd> or right-click -> **Rename Symbol**, it won't work:

![image](https://github.com/user-attachments/assets/a497f0f4-4d6b-4bf2-9187-d5527291787b)

This is likely because the rename action works if the caret is within the word, but not at the end of the word. However, selecting by double-clicking or <kbd>Ctrl + D</kbd> always puts the caret at the end of the word. Incindentally, it will currently work if there's a punctuation character or line break just after the symbol (but not if there's a space).

- The rename field should use the same code font as the script editor for consistency:

![image](https://github.com/user-attachments/assets/f2c66a6e-769c-4472-b696-9e4192de82bf)

- If you try to rename a symbol that can't be renamed (e.g. a method on a core engine type), a small popup will appear below the caret for a split second and the script editor will lose focus. Pressing <kbd>F2</kbd> again will rename the currently selected node in the Scene tree dock due to this:

https://github.com/user-attachments/assets/43f1c982-146e-4776-80c7-7d9953168a8f

- It's possible to create syntactically incorrect code by renaming a symbol to an invalid identifier, e.g. replacing `taken` with `taken 123` in this example. There should probably be a check for the value in the text field being a valid identifier, and disabling the Rename button if it's invalid.

- Undo/redo functionality doesn't seem to work correctly, even though history entries are created:

![image](https://github.com/user-attachments/assets/2bac3c7d-136d-4bca-a4b3-6f8fc4cda883)

If I press <kbd>Ctrl + Z</kbd>, nothing happens. However, selecting history entries will correctly go back to the previous state, providing the renamed value is a correct identifier (if not, words will be mangled around).

KoBeWi commented:
I did a test in a rather complex project. Renaming classes seems to work overall, but it does not rename type hints (as already reported above).

HOWEVER if you have 2 nested classes of the same name, e.g.
```GDScript
class Class1
	class Nested
		pass

class Class2
	class Nested
		pass
```
Renaming `Class1.Nested` will also rename `Class2.Nested`. It will however correctly rename references.

There is something wrong with renaming things in the current file. It works with classes, but seems to fail with enum names, members and constants (and probably more, but that's what I tested). If you try to rename them, they will get renamed correctly everywhere except the file where you started the rename.

The F2 shortcut does not work for me.

Also when renaming, lots of errors are spammed in the output.

> When using this in an unsaved file the changes get mixed up with the code from disk.

Still happening.

adamscott commented on editor/editor_node.cpp:
```diff
@@ -1588,6 +1591,25 @@ void EditorNode::save_resource_as(const Ref<Resource> &p_resource, const String
 	file->popup_file_dialog();
 }
 
+void EditorNode::save_resource_bulk(const LocalVector<Ref<Resource>> &p_resource_bulk) {
```
This is my attempt to reduce the save time. I need to check if it's still needed, though

adamscott commented on modules/gdscript/gdscript_editor.cpp:
```diff
@@ -28,12 +28,16 @@
 #include "gdscript_tokenizer.h"
 #include "gdscript_utility_functions.h"
+#include "modules/gdscript/gdscript_cache.h"
+#include <cstdint>
```
Oops. This shouldn't be here.

Mickeon commented on doc/classes/CodeEdit.xml:
```diff
@@ -604,6 +604,12 @@
 				Emitted when the user requests code completion. This signal will not be sent if [method _request_code_completion] is overridden or [member code_completion_enabled] is [code]false[/code].
 			</description>
 		</signal>
+		<signal name="refactor_requested">
+			<param index="0" name="type" type="int" />
+			<description>
+				Emitted when the user requests a refactor.
+			</description>
+		</signal>
```
Oh my god, this is particularly egregious. In its current state (prone to future changes), it should not be exposed at all, let alone without its corresponding enum. 

CC @KoBeWi because of https://github.com/godotengine/godot/pull/69012#discussion_r2051790554. We somehow still can't choose to keep signals unexposed.

KoBeWi replied:
CodeEdit has `code_completion_requested` too, so it's not completely out of place.

...if not the fact that this signal is only emitted externally.

kitbdev replied:
It looks like signals can be hidden from docs as of recently.
- https://github.com/godotengine/godot/pull/105691

Mickeon replied:
I'm not sure that means they're completely hidden from the API, though. Things like autocompletion may recommend them, the Node dock may display them. GDExtension may have them exposed anyway, though the leading underscore is a deterrent.

Mickeon replied:
My only concern with this PR is still this specific signal. @adamscott , can you consider a workaround as KoBeWi also discussed in https://github.com/godotengine/godot/pull/69012#discussion_r2051790554?
It would be an admittedly rare, but not unique case to use `Object.add_user_signal()` for this purpose.
https://github.com/godotengine/godot/blob/7864ac80192e9e91bf56176af9f04cc013b580aa/scene/2d/sprite_2d.cpp#L556
https://github.com/godotengine/godot/blob/7864ac80192e9e91bf56176af9f04cc013b580aa/scene/gui/code_edit.cpp#L3906

kitbdev replied:
It could also just be removed, like I suggest here https://github.com/godotengine/godot/pull/102380#discussion_r2099243262
There's no need for it in CodeEdit.

Mickeon commented on doc/classes/TextEdit.xml:
```diff
@@ -698,11 +698,21 @@
 				Returns the total number of lines between [param from_line] and [param to_line] (inclusive) in the text. This includes wrapped lines and excludes folded lines. If the range covers all lines it is equivalent to [method get_total_visible_line_count].
 			</description>
 		</method>
+		<method name="get_word_at_line_column" qualifiers="const">
+			<return type="String" />
+			<param index="0" name="line" type="int" />
+			<param index="1" name="column" type="int" />
+			<description>
+				Returns the word at the specified [param line] and [param column].
+				[b]Note:[/b] To get a word under a caret, use [method get_word_under_caret] instead.
+			</description>
+		</method>
```
Sounds useful, but hold on: Exposing it doesn't look necessary for this PR. It sounds like something that could be done separately upon request.

adamscott commented:
> * If you select a word by double-clicking or Ctrl + D then try to use F2 or right-click -> **Rename Symbol**, it won't work:

I couldn't reproduce. Weird.

> * The rename field should use the same code font as the script editor for consistency:

Done. I also added color to the text based on the symbol type!

![Capture d’écran du 2025-05-20 14-16-57](https://github.com/user-attachments/assets/f6ef3794-55e1-4a7b-8106-a9fa50ef1994)

> * If you try to rename a symbol that can't be renamed (e.g. a method on a core engine type), a small popup will appear below the caret for a split second and the script editor will lose focus. Pressing F2 again will rename the currently selected node in the Scene tree dock due to this:

Fixed! And it's even better now.

![Capture d’écran du 2025-05-20 14-16-49](https://github.com/user-attachments/assets/26794b1d-aae8-499d-b0f0-c4c930a8f09a)
![Capture d’écran du 2025-05-20 14-16-45](https://github.com/user-attachments/assets/d93a3ce1-cf7e-4e71-9d13-38e868b59bb1)
![Capture d’écran du 2025-05-20 14-16-10](https://github.com/user-attachments/assets/fe43b684-d84b-4794-beff-b10726a4e5c1)

kitbdev commented on editor/code_editor.h:
```diff
@@ -30,19 +30,80 @@
 class CodeTextEditor;
 class LineEdit;
 
+class RefactorRenamePopup : public PanelContainer {
```
Is there a reason that this isn't a Popup / PopupPanel?
The functionality is similar and it should be on top, so it doesn't get cut off:

![image](https://github.com/user-attachments/assets/8f25d84f-162e-4d57-ac81-a4db7e4bcdde)

adamscott replied:
Ah, I remember. If it's a popup, it doesn't follow the code scrolling. I think the box following its source is really important.

I'll try though to calculate the position of the fake popup better.

kitbdev reviewed (commented):
Only tested a little.

When trying to rename something in a comment or an unexpected identifier (with the shortcut) you get an error `ERROR: An error occurred during refactoring.`

I also had issues with unsaved changes reverting.

eimfach commented:
Is it possible with this to non-destructively rename exported variables ?

adamscott committed:
Fix issues with pre-commit

ManOfDuck commented:
> EDIT: Also renaming does not find references inside built-in scripts 🙃

Is this a planned feature of this PR? It would be incredibly useful but I could see it being out of scope.

uwu-420 commented:
Love to see this being worked on!

Will this be also available via LSP for example when using VS Code?

adamscott commented:
> Love to see this being worked on!
> 
> Will this be also available via LSP for example when using VS Code?

Currently and for some reason, the LSP and the GDScript editor tools don't share the same code. So edits to one unfortunately don't affect the other. Ideally, both would be merged together someday.

adamscott committed:
Set vformat content to one-line

adamscott committed:
Fix issue where argument index would be incremented twice

adamscott committed:
Add comment to reflect absence of refactor rename

adamscott committed:
Fix what seems to be an oversight

adamscott committed:
Rename `get_word_pos` to `get_word_pos_at_pos`

RobTheFiveNine commented:
It is a slightly off topic suggestion... but if the part of this change that identifies the instances of the symbol being renamed could be exposed via a "Find References" UI function, it would be great. The two functions [I find, at least] in other editors tend to be used in tandem, e.g. reviewing what things use a particular symbol before refactoring said symbol.

Thank you for working on this also - being able to rename symbols is one of the big things I feel I miss when using the built-in editor (alongside the ability to find all references to something easily), so this work is really appreciated 🎉

adamscott commented:
> It is a slightly off topic suggestion... but if the part of this change that identifies the instances of the symbol being renamed could be exposed via a "Find References" UI function, it would be great. The two functions [I find, at least] in other editors tend to be used in tandem, e.g. reviewing what things use a particular symbol before refactoring said symbol.
> 
> Thank you for working on this also - being able to rename symbols is one of the big things I feel I miss when using the built-in editor (alongside the ability to find all references to something easily), so this work is really appreciated 🎉

The idea of this PR is to integrate the functionality ASAP. The UI is intended to be implemented in a future PR, hopefully for 4.6 as well.

adamscott committed:
Better separate before/after

adamscott committed:
Add WIP of test cases for refactor

adamscott committed:
Remove commented code

adamscott committed:
WIP adding refactor test

adamscott committed:
Fix issue caused by godotengine/godot#107469

adamscott committed:
Add .global.gd extension to refactor global class files

adamscott committed:
Add refactor exception

adamscott committed:
Fix refactor test issues with global classes

adamscott committed:
Actually add a true test

adamscott committed:
Fix use of designated initializers

adamscott committed:
Add missing logic

adamscott committed:
Fix some detection issue and add new internal test

adamscott committed:
Fix variable shadowing

adamscott committed:
Fix issue where default font was not loaded

adamscott committed:
Add `load_scenes()`

adamscott committed:
Loop through scene state instead of instantiating

adamscott committed:
Add goto next and checks for .godot

adamscott committed:
Fix issue with existing extension signals

adamscott committed:
Huge overhaul of the testing

- Fix project setting root issue in all gdscript tests
  - Now, files are copied to a temp directory before testing
  - Each test has their own root
- Add refactor_rename_helper.py script to help finding matches

adamscott committed:
Remove superfluous file_access import

adamscott committed:
Add new output values for cfg to test if the refactor failed

adamscott commented on modules/gdscript/gdscript_editor.cpp:
```diff
@@ -3927,6 +4184,985 @@ void GDScriptLanguage::auto_indent_code(String &p_code, int p_from_line, int p_t
+							r_result.outside_refactor = true;
+							return OK;
+						} else {
+							const int dot_pos = enum_name.rfind_char('.');
+							if (dot_pos >= 0) {
+								r_result.type = ScriptLanguage::RefactorRenameSymbolResultType::REFACTOR_RENAME_SYMBOL_RESULT_CLASS_CONSTANT;
+								r_result.outside_refactor = true;
+								return OK;
```
Found a new issue while modifying an Enum value by referring to it by its outside classes (i.e. refactoring `ENUM_VALUE_1` from `InnerInnerB.EnumOfInnerInnerB.ENUM_VALUE_1`)

This is due to that code here.

adamscott committed:
Fix missing parsing_type default value

adamscott committed:
Add missing space in fail condition message

adamscott committed:
Add new test (will fail, need to fix issue)

adamscott committed:
Add check for new symbol if defined in test

adamscott committed:
Add separate project dir logic from `init_language()`

adamscott committed:
Move included files after `TOOLS_ENABLED`

adamscott committed:
Move refactor_rename_helper.py to misc/scripts

adamscott committed:
Remove superfluous newline

kitbdev commented on editor/script/script_text_editor.cpp:
```diff
@@ -1742,6 +1780,9 @@ void ScriptTextEditor::_edit_option(int p_op) {
 			tx->request_code_completion(true);
 		} break;
+		case EDIT_REFACTOR_RENAME_SYMBOL: {
+			tx->request_refactor(CodeEdit::RefactorKind::REFACTOR_KIND_RENAME);
```
CodeEdit has functionality for code completion, it draws and maintains the options and provides methods, settings, virtual methods, etc for it. Compared to this, CodeEdit has nothing to do with refactoring and it doesn't need to either. So I still think we should not affect CodeEdit if we can help it, especially if it means exposing a signal.

adamscott committed:
Remove refactor.to for the moment and fix match values

adamscott committed:
Add OUTSIDE_REFACTOR macro for rename_symbol_from_base

adamscott committed:
Fix shadowing issue

adamscott committed:
Fix text_server.h path

adamscott committed:
Replace old code with function call

adamscott committed:
Add `operator!=` to refactor matches

adamscott committed:
Fix enum matching issue

adamscott committed:
WIP to fix issue with enum refactor

KoBeWi commented:
I tried to test this again, but it doesn't seem to work at all currently. The rename option (both from menu and shortcut) just hangs editor for a couple seconds and nothing happens.

https://github.com/user-attachments/assets/0e9a085b-ad50-4e40-b023-4bfb5e306d8c

adamscott committed:
Add new add_match signature

adamscott committed:
Add deep parameter

adamscott committed:
Add missing info

adamscott committed:
Add new test and try to simplify symbol match loop

adamscott committed:
Simplify from class loop nodes logic

adamscott committed:
Make process_code return an error

adamscott committed:
Add member_variable rename tests

adamscott committed:
Fix issue with matchmap_to_string

adamscott committed:
Add two external script signal tests

adamscott committed:
Add inner script signal test

adamscott committed:
Fix issues with enum and signal matching

adamscott committed:
Fix font used

adamscott committed:
Fix shadowing of variable

adamscott committed:
Make GDScriptCache clearable multiple times

Fixes refactor tests memory leaks.

YeldhamDev commented:
The rename popup appears at the top of the script instead of below the symbol (on GNU/Linux at least).
<img width="437" height="197" alt="image" src="https://github.com/user-attachments/assets/bd24eb6c-55f7-4409-b985-51da719a3c4e" />

Calinou reviewed (commented):
Tested locally, it works as expected.

Some feedback:

- I can confirm the issue @YeldhamDev noticed with the incorrect popup position, regardless of whether **Single Window Mode** is enabled in the Editor Settings. I'm on Linux with X11.

- If you follow this scenario:
  - Rename an identifier to something invalid.
  - Press <kbd>Ctrl + Z</kbd> once to try to undo the change (it won't do anything, except move your caret).
  - Fix the script error manually.
  - Press <kbd>F2</kbd> while an identifier is selected (or when the caret is anywhere on an identifier).
  - Then it won't show the refactor/rename dialog until you restart the editor (reloading the scene or closing/reopening the script won't suffice). It hides the caret, and nothing happens if I type some text and press <kbd>Enter</kbd>.

- It's possible to enter invalid or reserved identifiers in the rename dialog (e.g. `something like this` or `Vector4`), which will cause script errors. Since the refactor currently can't be undone with <kbd>Ctrl + Z</kbd>, this can be quite damaging.
  - I suggest only allowing valid identifiers, and checking whether it's reserved (built-in engine type or keyword) before accepting the rename.
  - The changes from https://github.com/godotengine/godot/pull/102380#issuecomment-2895419274 no longer seem to be in this PR? The same goes for syntax highlighting, which no longer happens when entering known identifiers like user types.

- Leading and trailing whitespace should be trimmed from the string to rename to, using `String.strip_edges()`.

adamscott committed:
Fix refactor popup popping issues

adamscott committed:
Fix issues with local variables

adamscott committed:
Add additionnal local_variable internal test

adamscott committed:
Add lambda test for local variable rename

KoBeWi commented:
Renaming classes still doesn't work correctly.
You can test it here:
https://github.com/KoBeWi/Metroidvania-System
Rename e.g. `CellData` class from `MapData.gd`. It results in lots of broken references to the old name.

btw it would be nice if you squashed the commits, or at least avoided doing force-push. The long commit list buries all discussion :/

adamscott committed:
Add a test for renaming literals

adamscott committed:
Add test output for `refactor_result_type`

adamscott committed:
Add `refactor_result_type` to existing tests

adamscott committed:
Simplify `refactor_rename_symbol_code()` macros

adamscott committed:
Make use of SUBCASE (makes it easy to filter tests)

adamscott committed:
Add support to (try to) "rename" `void`

adamscott committed:
Add rename Node

adamscott committed:
Fix enum indent

adamscott committed:
WIP to get symbol directly from language

adamscott committed:
Fix isues with mismatch in extents tracking stack

adamscott committed:
Rename to `AwaitNode::keyword`

adamscott committed:
Add token in `GDScriptParser::Node`

adamscott committed:
Add missing empty.gd uid

adamscott committed:
Make refactor match on beginning of token

adamscott committed:
Add REFACTOR_RENAME_TYPE_CONTROL_FLOW

adamscott committed:
Add break, continue, and class_name tests

AThousandShips commented on scene/gui/code_edit.cpp:
```diff
@@ -32,6 +32,7 @@
 #include "code_edit.compat.inc"
 
 #include "core/config/project_settings.h"
+#include "core/object/object.h"
```
```suggestion
```
Already included

AThousandShips commented on editor/script/script_text_editor.cpp:
```diff
@@ -33,17 +33,22 @@
 #include "core/config/project_settings.h"
 #include "core/io/json.h"
 #include "core/math/expression.h"
+#include "core/object/script_language.h"
```
```suggestion
```

AThousandShips commented on editor/gui/code_editor.cpp:
```diff
@@ -30,19 +30,272 @@
 
 #include "code_editor.h"
 
+#include "core/error/error_macros.h"
```
```suggestion
```

AThousandShips commented on editor/gui/code_editor.cpp:
```diff
@@ -30,19 +30,272 @@
+#include "scene/gui/panel_container.h"
 #include "scene/gui/separator.h"
+#include "scene/main/node.h"
+#include "scene/main/window.h"
```
```suggestion
```

AThousandShips commented on modules/gdscript/gdscript_editor.cpp:
```diff
@@ -28,12 +28,16 @@
 /**************************************************************************/
 
+#include "core/error/error_list.h"
+#include "core/string/print_string.h"
```
```suggestion
```

AThousandShips commented on scene/gui/text_edit.cpp:
```diff
@@ -29,6 +29,7 @@
 /**************************************************************************/
 
 #include "text_edit.h"
+#include "core/error/error_macros.h"
```
```suggestion
```

AThousandShips commented on editor/script/script_text_editor.cpp:
```diff
@@ -34,22 +34,27 @@
+#include "scene/gui/code_edit.h"
 #include "scene/gui/grid_container.h"
 #include "scene/gui/menu_button.h"
+#include "scene/gui/popup_menu.h"
```
```suggestion
```

AThousandShips commented on editor/gui/code_editor.h:
```diff
@@ -197,6 +260,16 @@ class CodeTextEditor : public VBoxContainer {
+	void _on_refactor_rename_popup_closed();
+	void _on_refactor_rename_popup_apply();
+	void _on_refactor_rename_popup_preview();
+	void _on_refactor_rename_popup_restore_caret(Point2i p_caret_position);
```
```suggestion
	void _on_refactor_rename_popup_restore_caret(const Point2i &p_caret_position);
```

AThousandShips commented on editor/gui/code_editor.h:
```diff
@@ -31,19 +31,79 @@
+	Point2i get_code_position();
+	ScriptLanguage::RefactorRenameSymbolResult get_refactor_result();
+
+	void request_refactor(const ScriptLanguage::RefactorRenameSymbolResult &p_refactor_result, Point2i p_code_position, Point2i p_caret_position, Point2i p_selection_start_position = Point2i(-1, -1));
```
```suggestion
	void request_refactor(const ScriptLanguage::RefactorRenameSymbolResult &p_refactor_result, const Point2i &p_code_position, const Point2i &p_caret_position, const Point2i &p_selection_start_position = Point2i(-1, -1));
```

AThousandShips commented on editor/gui/code_editor.cpp:
```diff
@@ -1093,6 +1346,277 @@ Ref<Texture2D> CodeTextEditor::_get_completion_icon(const ScriptLanguage::CodeCo
+				int i = 0;
+				while (i < actual_character_index) {
+					String character = lines[target_line].substr(i, 1);
+					if (character == String("\t")) {
```
```suggestion
					if (character == "\t") {
```

AThousandShips commented on modules/gdscript/gdscript_parser.cpp:
```diff
@@ -68,6 +68,38 @@ Variant::Type GDScriptParser::get_builtin_type(const StringName &p_type) {
 }
 
+// This functions returns the position of the cursor sentinel from a given source code.
+Vector2i GDScriptParser::get_cursor_sentinel_position(const String &p_source_code, int tab_size) {
```
```suggestion
Vector2i GDScriptParser::get_cursor_sentinel_position(const String &p_source_code, int p_tab_size) {
```

AThousandShips commented on editor/gui/code_editor.cpp:
```diff
@@ -30,19 +30,340 @@
+	// rename_container->add_child(preview_button);
+	// preview_button->set_text(TTR("Preview"));
+	// preview_button->connect(SNAME("pressed"), callable_mp(this, &RefactorRenamePopup::_on_preview_button_pressed));
+	// preview_button->connect(SceneStringName(focus_exited), callable_mp(this, &RefactorRenamePopup::_on_focus_exited));
```
```suggestion
	//preview_button = memnew(Button);
	//rename_container->add_child(preview_button);
	//preview_button->set_text(TTR("Preview"));
	//preview_button->connect(SceneStringname(pressed), callable_mp(this, &RefactorRenamePopup::_on_preview_button_pressed));
	//preview_button->connect(SceneStringName(focus_exited), callable_mp(this, &RefactorRenamePopup::_on_focus_exited));
```
Or just:
```suggestion
```

AThousandShips commented on editor/gui/code_editor.h:
```diff
@@ -30,9 +30,16 @@
+#include "scene/gui/panel_container.h"
+#include "scene/gui/popup.h"
+#include "scene/gui/rich_text_label.h"
+#include "scene/main/timer.h"
```
Remove all these forward declared types

AThousandShips commented on editor/gui/code_editor.cpp:
```diff
@@ -30,23 +30,343 @@
+	return !symbol_edit_text.is_empty();
+}
+
+void RefactorRenamePopup::set_state(RefactorRenamePopup::State p_state) {
```
```suggestion
void RefactorRenamePopup::set_state(State p_state) {
```

adamscott committed:
Add annotation support and test

AThousandShips commented on modules/gdscript/tests/gdscript_test_runner.cpp:
```diff
@@ -110,15 +112,71 @@ void init_autoloads() {
+
+#define COPY_DIR(from, to)        \
+	err = da->copy_dir(from, to); \
+	CHECK_MESSAGE(err == OK, vformat("error while copying dir \"%s\" to \"%s\"", from, to))
```
```suggestion
#define COPY_DIR(m_from, m_to)        \
	err = da->copy_dir(m_from, m_to); \
	CHECK_MESSAGE(err == OK, vformat("Error while copying dir \"%s\" to \"%s\".", m_from, m_to))
```

AThousandShips commented on modules/gdscript/gdscript_editor.cpp:
```diff
@@ -3927,6 +4184,883 @@ void GDScriptLanguage::auto_indent_code(String &p_code, int p_from_line, int p_t
+static Error _refactor_rename_symbol_from_base(GDScriptParser::RefactorRenameContext p_context, const GDScriptParser::DataType &p_base, const String &p_symbol, const String &p_path, const String &p_class_path, Object *p_owner, const HashMap<String, String> &p_unsaved_scripts_source_code, ScriptLanguage::RefactorRenameSymbolResult &r_result) {
+#define OUTSIDE_REFACTOR(_type)                                            \
+	r_result.type = ScriptLanguage::RefactorRenameSymbolResultType::_type; \
+	r_result.outside_refactor = true;                                      \
```
```suggestion
#define OUTSIDE_REFACTOR(m_type)                                            \
	r_result.type = ScriptLanguage::RefactorRenameSymbolResultType::m_type; \
	r_result.outside_refactor = true;                                       \
```

AThousandShips commented on modules/gdscript/tests/test_refactor.h:
```diff
@@ -0,0 +1,470 @@
+		bool refactor_has_failed = conf.get_value("output", "refactor_failed", false);
+
+		Ref<GDScript> gdscript = ResourceLoader::load(refactor_file);
+		FAIL_COND_MSG(gdscript.is_null(), vformat("couldn't load script \"%s\"", refactor_file));
```
```suggestion
		FAIL_COND_MSG(gdscript.is_null(), vformat(R"(Couldn't load script "%s".)", refactor_file));
```

AThousandShips commented on modules/gdscript/tests/test_refactor.h:
```diff
@@ -0,0 +1,470 @@
+			LocalVector<Match> refactor_result_key_matches = KV.value;
+
+			if (expected_result_matches.size() > 0) {
+				CHECK_MESSAGE(expected_result_matches.has(refactor_result_key), vformat("unexpected match for \"%s\" not defined in the cfg", refactor_result_key));
```
```suggestion
				CHECK_MESSAGE(expected_result_matches.has(refactor_result_key), vformat(R"(Unexpected match for "%s" not defined in the cfg.)", refactor_result_key));
```

AThousandShips commented on modules/gdscript/tests/gdscript_test_runner.cpp:
```diff
@@ -110,15 +112,71 @@ void init_autoloads() {
+	// Initialize res:// to `project_folder`.
+	TestProjectSettingsInternalsAccessor::resource_path() = project_folder;
+	err = ProjectSettings::get_singleton()->setup(project_folder, String(), true);
+	CHECK_MESSAGE(err == OK, "error while setuping project");
```
```suggestion
	CHECK_MESSAGE(err == OK, "Error while setting up project.");
```

AThousandShips commented on modules/gdscript/gdscript_editor.cpp:
```diff
@@ -3927,6 +4184,882 @@ void GDScriptLanguage::auto_indent_code(String &p_code, int p_from_line, int p_t
+								continue;
+							}
+						}
+						ERR_FAIL_COND_V_MSG(identifier_node_owner == nullptr, FAILED, "identifier node owner datatype is null");
```
```suggestion
						ERR_FAIL_NULL_V_MSG(identifier_node_owner, FAILED, "Identifier node owner datatype is null.");
```

AThousandShips commented on modules/gdscript/tests/test_refactor.h:
```diff
@@ -0,0 +1,536 @@
+		Error refactor_error_result = GDScriptLanguage::get_singleton()->refactor_rename_symbol_code(code_with_sentinel, refactor_file, owner, unsaved_scripts_code, refactor_result);
+		FAIL_COND_MSG(refactor_error_result != OK, vformat("could not refactor rename symbol code: %s", error_names[refactor_error_result]));
+
+		FAIL_COND_MSG(refactor_result.symbol != refactor_symbol, vformat("symbol found (%s) doesn't match with the symbol specified in the configuration file (%s)", refactor_result.symbol, refactor_symbol));
```
```suggestion
		FAIL_COND_MSG(refactor_result.symbol != refactor_symbol, vformat("Symbol found (%s) doesn't match with the symbol specified in the configuration file (%s).", refactor_result.symbol, refactor_symbol));
```

adamscott committed:
WIP for adding tokens in `GDScriptParser::Node`

adamscott committed:
Add tokens to parse_grouping()

adamscott committed:
Add token to `parse_cast()`

adamscott committed:
Add token to `parse_call()`

adamscott committed:
Add tokens to `parse_get_node()` and `parse_preload()`

adamscott committed:
Remove commented code

adamscott committed:
Parse successfully `{Array,Break,Continue}Node` tokens

adamscott committed:
Remove tokens in `GDScriptParser::Node`

- Add `GDScriptTokenizer::CodeArea`
- Add `GDScriptParser::get_token(int p_line, int p_column)`

adamscott committed:
Cleanup registration of refactor context

adamscott committed:
Fix `CodeArea::*_or_together` methods

adamscott committed:
Remove null error from `GDScriptParser::refactor_rename_is_node_more_specific()`

Add check for `refactor_rename_context.node == nullptr`

adamscott committed:
Fix GDScriptTokenizer::CodeArea::contains(const Pair<int,int>)

- This add granular overlap checks for start or end

adamscott committed:
Fix variable parsing

adamscott committed:
Fix refactor context token fetching code

adamscott committed:
Fix function refactor

adamscott committed:
Fix renaming "Node"

adamscott committed:
Add rename_extends test

adamscott committed:
Fix renaming void

adamscott committed:
Fix rename class_name

adamscott committed:
Fix tests/type/rename_int.cfg

adamscott committed:
Fix tests/variable/local_variable/lambda.cfg

adamscott committed:
Fix tests/signal/external_script.cfg

adamscott committed:
WIP to rethink the context concept

adamscott committed:
Remove literal as refactor context

adamscott committed:
Fix parse_type()

adamscott committed:
Fix "extends" issue.

adamscott committed:
Fix rename string line issue

AThousandShips commented on modules/gdscript/gdscript_tokenizer.h:
```diff
@@ -44,6 +44,314 @@ class GDScriptTokenizer {
+		constexpr bool is_overlapping(Vector2i p_position) const {
+			return is_overlapping(Pair<int, int>{ p_position.y, p_position.x });
+		}
+		constexpr bool is_overlapping(Vector2i p_start, Vector2i p_end) const {
```
```suggestion
		constexpr bool is_overlapping(const Vector2i &p_start, const Vector2i &p_end) const {
```
For all

AThousandShips commented on modules/gdscript/gdscript_parser.cpp:
```diff
@@ -89,6 +122,24 @@ bool GDScriptParser::annotation_exists(const String &p_annotation_name) const {
+			return token;
+		}
+		if (token_area.is_after(position)) {
+			return {};
```
```suggestion
			return Token();
```

adamscott committed:
Add new `LineColumn` type (`Pair<int, int>`)

adamscott committed:
Fix issues with enum (detects correctly enum names/values)

AThousandShips commented on modules/gdscript/gdscript_tokenizer.h:
```diff
@@ -44,6 +44,316 @@ class GDScriptTokenizer {
+		}
+
+		constexpr CodeArea() = default;
+		constexpr CodeArea(LineColumn p_start, LineColumn p_end) :
```
```suggestion
		constexpr CodeArea(const LineColumn &p_start, const LineColumn &p_end) :
```

adamscott committed:
Remove superfluous print line

adamscott committed:
Fix attribute and enum matching

adamscott committed:
Move test into subtest dir

adamscott committed:
Fix signal refactor

adamscott committed:
Add the concept of `GDScriptParser::Node::owner`

adamscott committed:
Add missing `GDScriptTokenizer::Token::operator!=()`

adamscott committed:
Fix loop_nodes

adamscott committed:
Fix signals

adamscott committed:
Fix enums

adamscott committed:
Fix standalone literals and identifier statements

adamscott committed:
Fix await and precise the use of `REFACTOR_RENAME_TYPE_STATEMENT`

- Remove `REFACTOR_RENAME_TYPE_{IDENTIFIER,LITERAL}`

[no ci]

adamscott committed:
WIP (fix numerous issues)

[no ci]

adamscott committed:
Fix renaming class name

[no ci]

adamscott committed:
Fix array and dictionary

[no ci]

adamscott committed:
WIP

pirey0 commented:
Hey, maybe I'm missing something, but why are we not doing this through the LSP?
